src/qililab/qprogram/qblox_compiler.py

```diff
@@ -669,3 +713,30 @@ def _hash_waveform(waveform: Waveform):
             key: (value.__dict__ if isinstance(value, Waveform) else value) for key, value in waveform.__dict__.items()
         }
         return f"{waveform.__class__.__name__} {hashes}"
+
+    @staticmethod
+    def calculate_square_waveform_optimization_values(duration):
+        def remainder_conditions(pd):
+            rem = duration % pd
+            return rem, (pd >= 4 and (rem == 0 or rem >= 4))
+
+        def find_piece_duration(condition_func):
+            for pd in range(100, 501):
+                if pd <= duration:
+                    rem, valid = remainder_conditions(pd)
+                    if valid and condition_func(rem):
+                        return pd
+            return None
+
+        # First try for remainder == 0
+        final_pd = find_piece_duration(lambda rem: rem == 0)
+        if final_pd is not None:
+            return final_pd, duration // final_pd, duration % final_pd
+
+        # If not found, try for remainder ≥ 4
+        final_pd = find_piece_duration(lambda rem: rem >= 4)
+        if final_pd is not None:
+            return final_pd, duration // final_pd, duration % final_pd
+
+        # If no suitable piece_duration found, fallback to entire duration
+        return duration, 1, 0
```

Comment:
Do you think is worthy raising a warning here telling the user it tried to optimise for values that do not allow optimisation?

Reply:
not really, this is just a fallback e.g. for small waveforms like 40ns that will stay the same.

Reply:
Cool, good on my side once changelog and codecoverage have been sorted. good job.